psutil/arch/osx/cpu.c

```diff
@@ -109,7 +114,94 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
     );
 }
 
+#if defined(__arm64__) || defined(__aarch64__)
+PyObject *
+psutil_cpu_freq(PyObject *self, PyObject *args) {
+    uint32_t min;
+    uint32_t curr;
+    uint32_t pMin;
+    uint32_t eMin;
+    uint32_t max;
+    kern_return_t status;
+    CFDictionaryRef matching = NULL;
+    CFTypeRef pCoreRef = NULL;
+    CFTypeRef eCoreRef = NULL;
+    io_iterator_t iter = 0;
+    io_registry_entry_t entry = 0;
+
+    matching = IOServiceMatching("AppleARMIODevice");
+    if (matching == 0) {
+        return PyErr_Format(PyExc_RuntimeError, "IOServiceMatching call failed, 'AppleARMIODevice' not found");
+    }
+
+    status = IOServiceGetMatchingServices(kIOMainPortDefault, matching, &iter);
+    if (status != KERN_SUCCESS) {
+        PyErr_Format(PyExc_RuntimeError, "IOServiceGetMatchingServices call failed");
+        goto error;
+    }
 
+    while ((entry = IOIteratorNext(iter))) {
+        io_name_t name;
+        status = IORegistryEntryGetName(entry, name);
+        if (status != KERN_SUCCESS) {
+            if (entry)
+                IOObjectRelease(entry);
+            continue;
+        }
+        if (strcmp(name, "pmgr") == 0) {
+            break;
+        }
+        IOObjectRelease(entry);
+    }
+
+    if (entry == 0) {
+        PyErr_Format(PyExc_RuntimeError, "'pmgr' entry was not found in AppleARMIODevice service");
+        goto error;
+    }
+
+    pCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states5-sram"), kCFAllocatorDefault, 0);
+    if (!pCoreRef) {
+        PyErr_Format(PyExc_RuntimeError, "'voltage-states5-sram' property not found");
+        goto error;
+    }
+
+    eCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states1-sram"), kCFAllocatorDefault, 0);
+    if (!eCoreRef) {
+        PyErr_Format(PyExc_RuntimeError, "'voltage-states1-sram' property not found");
+        goto error;
+    }
+
```

Comment:
You've released all the `entry` instances in the loop and in error messages, but you need one more here. :-)